### PR TITLE
feat: High performance pandas integration.

### PR DESCRIPTION
**Pull request by amunra**
* High-performance ingestion of [Pandas](https://pandas.pydata.org/) dataframes into QuestDB via ILP.
  We now support most Pandas column types. The logic is implemented in native code and is orders of magnitude faster than iterating the dataframe in Python and calling the `Buffer.row()` or `Sender.row()` methods: The `Buffer` can be written from Pandas at hundreds of MiB/s per CPU core. The new `dataframe()` method continues working with the `auto_flush` feature.
* New `TimestampNanos.now()` and `TimestampMicros.now()` methods. *These are the new recommended way of getting the current timestamp.*
* The Python GIL is now released during calls to `Sender.flush()` and when `auto_flush` is triggered. This should improve throughput when using the `Sender` from multiple threads.
* General documentation clean-up.

### CODE REVIEW

File: pystr-to-utf8/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    let b = &mut *b;
+    b.0.truncate(pos.chain);
+    if !b.0.is_empty() {
+        b.0[pos.chain - 1].truncate(pos.string);
+    }
+}
+
+/// Reset the converter's buffer to zero length.
+#[no_mangle]
+pub unsafe extern "C" fn qdb_pystr_buf_clear(b: *mut qdb_pystr_buf) {
```
[the remark this answers is not in the document]
Comment (amunra) [reply pystr-to-utf8/src/lib.rs:73]: True. It was my original thought. The `Vec::clear` codepath however is a fair bit simpler in its implementation.
Would you mind if I keep it as-is?

File: pystr-to-utf8/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            0
+        };
+    qdb_pystr_pos { chain: chain_pos, string: string_pos }
+}
+
+/// Trim the buffer to the given position. Use in conjunction with `tell`.
+#[no_mangle]
+pub unsafe extern "C" fn qdb_pystr_buf_truncate(
+        b: *mut qdb_pystr_buf, pos: qdb_pystr_pos) {
+    let b = &mut *b;
```
[the remark this answers is not in the document]
Comment (amunra) [reply pystr-to-utf8/src/lib.rs:64]: I've checked and there isn't a single codepath that sets a `qdb_pystr_buf*` to a `NULL`.
No point in such defensive programming for a non-public API.

File: pystr-to-utf8/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+const MIN_BUF_LEN: usize = 1024;
+
+/// A carefully crafted buffer with spare capacity for `len` bytes.
+/// This is necessary to return "stable" addresses and avoid segfaults.
+/// Rust is unaware we are borrowing its memory and could try to free it as
+/// part of a reallocation if we were to use a `String` directly.
+fn get_dest<'a>(chain: &'a mut Vec<String>, len: usize) -> &'a mut String {
+    if !chain.is_empty() {
+        let last = chain.last_mut().unwrap();
```
[the remark this answers is not in the document]
Comment (amunra) [reply pystr-to-utf8/src/lib.rs:97]: Yeah, annoying.. isn't it.

File: pystr-to-utf8/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    }
+}
+
+const MIN_BUF_LEN: usize = 1024;
+
+/// A carefully crafted buffer with spare capacity for `len` bytes.
+/// This is necessary to return "stable" addresses and avoid segfaults.
+/// Rust is unaware we are borrowing its memory and could try to free it as
+/// part of a reallocation if we were to use a `String` directly.
+fn get_dest<'a>(chain: &'a mut Vec<String>, len: usize) -> &'a mut String {
```
[the remark this answers is not in the document]
Comment (amunra) [reply pystr-to-utf8/src/lib.rs:95]: Nice one!
Done.

File: src/questdb/pandas_integration.md [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+1     <NA>
+2      str
+3  example
+>>> df.dtypes['a']
+string[pyarrow]
+>>> type(df.dtypes['a']).mro()
+[<class 'pandas.core.arrays.string_.StringDtype'>, <class 'pandas.core.dtypes.base.StorageExtensionDtype'>, <class 'pandas.core.dtypes.base.ExtensionDtype'>, <class 'object'>]
+```
+
+Note that these strings will always have indicies based on `int32_t`.
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/pandas_integration.md:456]: Thanks.
Done.

File: src/questdb/pandas_integration.md [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+2      str
+3  example
+>>> df.dtypes['a']
+string[python]
+>>> type(df.dtypes['a']).mro()
+[<class 'pandas.core.arrays.string_.StringDtype'>, <class 'pandas.core.dtypes.base.StorageExtensionDtype'>, <class 'pandas.core.dtypes.base.ExtensionDtype'>, <class 'object'>]
+```
+
+Note that by default the storage is still Python objects (sigh),
+so our Rust-based conversion will will come handy here as well.
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/pandas_integration.md:419]: Thanks.
Done.

File: src/questdb/pandas_integration.md [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+1    1.0
+2    1.5
+3    2.0
+Name: c, dtype: float64
+```
+
+#### Arrow floats
+
+Pandas also has arrow-compatible floats.
+These have an additiona bitvector to represent nulls.
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/pandas_integration.md:271]: Thanks.
Done.

File: src/questdb/pandas_integration.md [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+untestability.
+
+To tame this and maintain one's sanity we need to remember that we
+don't need to support every type, data-structure et cetera that pandas, numpy
+and arrow can throw at us: Instead we approach this by only accepting
+one-dimensional arrays that support our basic ILP supported types _only_.
+
+We can also further simplify iteration via the introduction of a cursor:
+a struct that is a mishmash of the simplified subsets of arrow and py buffers
+that we we actually care about.
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/pandas_integration.md:46]: Thanks,
Done.

**Review by amyshwang — COMMENTED** [review]
Very minor suggestions

File: src/questdb/ingress.pyx [hunk 9]
```diff
@@ -764,181 +957,259 @@ cdef class Buffer:
...
+
+            * **δ**: Columns containing strings can also be used to specify the
+              table name. See ``table_name_col``.
+
+            * **ε**: We only support categories containing strings. If the
+              category contains non-string values, an error will be raised.
+
+            * **ζ**: We don't support datetimes with precisions other than
+              nanoseconds. The designated timestamp column (see ``at``
+              parameter) maintains the nanosecond precision, whilst values
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/ingress.pyx:1194]: Thanks. Done.

File: src/questdb/ingress.pyx [hunk 10]
```diff
@@ -764,181 +957,259 @@ cdef class Buffer:
...
+              stored as columns have their precision truncated to microseconds.
+              All dates are sent as UTC and any additional timezone information
+              is dropped. If no timezone is specified, we follow
+              the pandas convention of assuming the timezone is UTC.
+              Date times before 1970-01-01 00:00:00 UTC are not supported.
+              If a datetime value is specified as ``None`` (``NaT``), it is
+              interpreted as the current time set by the server on receipt of
+              message.
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/questdb/ingress.pyx:1202]: Done. Thanks!

File: dev_requirements.txt [hunk 11]
```diff
@@ -1,8 +1,10 @@
...
 Cython>=0.29.32
 wheel>=0.34.2
-cibuildwheel>=2.11.1
+cibuildwheel>=2.11.2
 Sphinx>=5.0.2
 sphinx-rtd-theme>=1.0.0
 twine>=4.0.1
 bump2version>=1.0.1
+pandas>=1.3.5
+numpy>=1.21.6
```
Comment (eugenels) [inline dev_requirements.txt:10]: pyarrow and fastparquet dependencies may be missing

Comment (amunra) [reply dev_requirements.txt]: Done.

File: docs/installation.rst [hunk 12]
```diff
@@ -5,9 +5,9 @@ Installation
 The Python QuestDB client does not have any additional run-time dependencies and
 will run on any version of Python >= 3.7 on most platforms and architectures.
 
-You can install it globally by running::
+You can install it (or updated it) globally by running::
```
[the remark this answers is not in the document]
Comment (amunra) [reply docs/installation.rst]: Thanks! Done.

**Review by eugenels — APPROVED** [review]

**Pull request merged**